tethys_portal/dependencies.py

```diff
@@ -0,0 +1,245 @@
+import logging
+
+from django.conf import settings
+from django.templatetags.static import static
+
+logger = logging.getLogger('tethys.' + __name__)
+
+
+class StaticDependency:
+    cdn_url = None
+    static_url = '/{npm_name}/{path}'
+
+    def __init__(self, npm_name, version, css_path=None, js_path=None, css_integrity=None, js_integrity=None,
+                 js_paths=None, debug_path_converter=None, cdn_url=None):
+        self.npm_name = npm_name
+        self.version = version
+        self.css_path = css_path
+        self.js_path = js_path
+        self.js_paths = js_paths
+        self.css_integrity = css_integrity
+        self.js_integrity = js_integrity
+        self.debug_path_converter = debug_path_converter or self.default_debug_path_converter
+        self.cdn_url = cdn_url or self.cdn_url
+
+    @property
+    def use_cdn(self):
+        return settings.STATICFILES_USE_CDN
+
+    @property
+    def css_url(self):
+        return self._get_url(url_type_or_path='css')
+
+    @property
+    def js_url(self):
+        return self._get_url(url_type_or_path='js')
+
+    @property
+    def link_tag(self):
+        return self._get_tag('css')
+
+    @property
+    def script_tag(self):
+        return self._get_tag('js')
+
+    @staticmethod
+    def default_debug_path_converter(path):
+        return path.replace('min.', '')
+
+    def get_custom_version_url(self, url_type, version, debug=None):
+        if version == self.version:
+            pass
+        elif not self.use_cdn:
+            logger.warning(
+                f'The "STATICFILES_USE_CDN" setting is set to False, so the custom version "{self.npm_name}={version}" '
+                f'is not supported. Using the default version ({self.version}) instead.'
```

Comment:
Why are custom versions not supported for npm?

Reply:
Only one version of the npm package is downloaded (the versions specified in the `vendor_static_dependencies` dict below. If we wanted to support custom versions of something then the user would have to manually download that version of the dependency. It may work, but would probably require some testing.

Reply:
I changed the behavior so that if a custom version is requested  when the `USE_NPM` option is `True` then a warning is issued saying it will attempt to use the CDN.